File: app/macos/hyperwhisper/Managers/Transcription/Support/PhoneticVocabularyMatcher.swift
//
//  PhoneticVocabularyMatcher.swift
//  hyperwhisper
//
//  PHONETIC VOCABULARY MATCHER
//  Uses the Beider-Morse phonetic algorithm to match misrecognized words
//  in Parakeet transcription output to user-defined vocabulary entries.
//
//  This catches phonetically similar errors that exact string matching misses,
//  e.g. "hyper wisper" → "HyperWhisper", "pair a keet" → "Parakeet".
//
//  Based on the approach used by Murmure (github.com/Kieirra/murmure).
//

import Foundation
import os

/// Matches transcription words against user vocabulary using phonetic similarity.
class PhoneticVocabularyMatcher {

    private let logger = Logger(subsystem: "com.hyperwhisper.app", category: "PhoneticVocabularyMatcher")

    /// A pre-encoded vocabulary entry for phonetic comparison.
    private struct EncodedEntry {
        let originalWord: String
        let phoneticCodes: [String]
    }

    /// Pre-encoded vocabulary for fast matching.
    private var encodedVocabulary: [EncodedEntry] = []

    /// Initialize the matcher with vocabulary items.
    /// Pre-encodes all vocabulary words phonetically for efficient matching.
    ///
    /// - Parameter vocabulary: Vocabulary items to match against.
    ///   Items without a replacement are used as-is (the word itself is the correct spelling).
    ///   Items with a replacement are ignored here (handled by VocabularyProcessor's regex replacement).
    init(vocabulary: [Vocabulary]) {
        for item in vocabulary {
            guard let word = item.word?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !word.isEmpty else { continue }

            // Skip items that have explicit replacements — those are handled by regex replacement
            if let replacement = item.replacement, !replacement.isEmpty {
                continue
            }

            // Skip very short words (≤2 chars) to avoid false positives
            guard word.count > 2 else { continue }

            let codes = BeiderMorse.encode(word)
            guard !codes.isEmpty else { continue }

            encodedVocabulary.append(EncodedEntry(originalWord: word, phoneticCodes: codes))
        }

        if !encodedVocabulary.isEmpty {
            logger.info("Phonetic matcher initialized with \(self.encodedVocabulary.count) vocabulary entries")
        }
    }

    /// Apply phonetic vocabulary matching to transcribed text.
    /// For each word in the transcription, checks if it phonetically matches a vocabulary entry.
    /// If a match is found, replaces the transcribed word with the correct vocabulary spelling.
    ///
    /// - Parameter text: Raw transcription text from Parakeet.
    /// - Returns: Text with phonetically matched vocabulary corrections applied.
    func apply(to text: String) -> String {
        guard !encodedVocabulary.isEmpty else { return text }

        var corrected = text
        let words = text.components(separatedBy: .whitespaces).filter { !$0.isEmpty }

        for word in words {
            // Skip very short words to avoid false positives
            guard word.count > 2 else { continue }

            // Strip trailing punctuation for matching; the replacement preserves the
            // original punctuation via the \b-anchored regex below (it sits outside the
            // word characters), so no manual re-attachment is needed.
            let (cleanWord, _) = stripTrailingPunctuation(word)
            guard !cleanWord.isEmpty else { continue }

            let candidateCodes = BeiderMorse.encode(cleanWord)
            guard !candidateCodes.isEmpty else { continue }

            // Check against each vocabulary entry
            for entry in encodedVocabulary {
                // Skip if the word already matches the vocabulary entry exactly (case-insensitive)
                if cleanWord.caseInsensitiveCompare(entry.originalWord) == .orderedSame {
                    break
                }

                // Check if any phonetic code matches
                let hasMatch = entry.phoneticCodes.contains { dictCode in
                    candidateCodes.contains(dictCode)
                }

                if hasMatch {
                    // Word-boundary anchored, escaped regex replace (matches VocabularyProcessor).
                    // Replace only the cleanWord that actually matched so substrings of other
                    // words are left intact (e.g. "Kat" no longer mangles "category"/"scatter").
                    // The \b sits between the word chars and any trailing punctuation, so the
                    // source punctuation survives untouched.
                    let pattern = "\\b\(NSRegularExpression.escapedPattern(for: cleanWord))\\b"
                    corrected = corrected.replacingOccurrences(
                        of: pattern,
                        with: NSRegularExpression.escapedTemplate(for: entry.originalWord),
                        options: [.regularExpression, .caseInsensitive]
                    )
                    logger.debug("Phonetic match: '\(word, privacy: .public)' → '\(entry.originalWord, privacy: .public)'")
                    break // Use first match
                }
            }
        }

        return corrected
    }

    /// Strip trailing punctuation from a word, returning the clean word and the punctuation.
    private func stripTrailingPunctuation(_ word: String) -> (String, String) {
        var clean = word
        var punctuation = ""
        while let last = clean.last, last.isPunctuation {
            punctuation = String(last) + punctuation
            clean.removeLast()
        }
        return (clean, punctuation)
    }
}

File: app/windows/HyperWhisper/Services/PhoneticVocabularyMatcher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using HyperWhisper.Data.Entities;
using uniffi.hyperwhisper_core;

namespace HyperWhisper.Services;

/// <summary>
/// PHONETIC VOCABULARY MATCHER
///
/// Uses the Beider-Morse phonetic algorithm (via the shared Rust core,
/// <c>hw-core</c>) to match misrecognized words in local transcription output to
/// user-defined vocabulary entries. Catches phonetically similar errors that
/// exact string matching misses, e.g. "hyper wisper" → "HyperWhisper".
///
/// Direct port of the macOS <c>PhoneticVocabularyMatcher.swift</c>; both
/// platforms now call the same <c>phonetic_encode</c> export, so behaviour stays
/// in lockstep. This capability is new on Windows.
///
/// Entries WITH a replacement are skipped here — those are handled by
/// <see cref="VocabularyProcessor"/>'s regex replacement.
/// </summary>
public sealed class PhoneticVocabularyMatcher
{
    private readonly record struct EncodedEntry(string OriginalWord, IReadOnlyList<string> PhoneticCodes);

    private readonly List<EncodedEntry> _encodedVocabulary = new();

    /// <summary>
    /// Build a matcher from vocabulary items, pre-encoding each word phonetically.
    /// </summary>
    /// <param name="vocabulary">
    /// Vocabulary items to match against. Items with an explicit replacement, or
    /// words of 2 characters or fewer, are ignored to avoid false positives.
    /// </param>
    public PhoneticVocabularyMatcher(IEnumerable<VocabularyItem> vocabulary)
    {
        if (vocabulary is null)
        {
            throw new ArgumentException("Vocabulary must not be null.", nameof(vocabulary));
        }

        foreach (var item in vocabulary)
        {
            var word = item.Word?.Trim();
            if (string.IsNullOrEmpty(word))
            {
                continue;
            }

            // Items with explicit replacements are handled by regex replacement.
            if (!string.IsNullOrEmpty(item.Replacement))
            {
                continue;
            }

            // Skip very short words (<= 2 chars) to avoid false positives.
            if (word.Length <= 2)
            {
                continue;
            }

            var codes = BeiderMorse.Encode(word);
            if (codes.Count == 0)
            {
                continue;
            }

            _encodedVocabulary.Add(new EncodedEntry(word, codes));
        }

        if (_encodedVocabulary.Count > 0)
        {
            LoggingService.Info($"Phonetic matcher initialized with {_encodedVocabulary.Count} vocabulary entries");
        }
    }

    /// <summary>
    /// Apply phonetic vocabulary matching to transcribed text. For each word,
    /// checks if it phonetically matches a vocabulary entry and, if so, replaces
    /// the transcribed word with the correct vocabulary spelling.
    /// </summary>
    public string Apply(string text)
    {
        if (_encodedVocabulary.Count == 0 || string.IsNullOrEmpty(text))
        {
            return text;
        }

        var corrected = text;
        var words = text.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);

        foreach (var word in words)
        {
            // Skip very short words to avoid false positives.
            if (word.Length <= 2)
            {
                continue;
            }

            // Strip trailing punctuation for matching; the \b-anchored regex
            // replacement preserves the original punctuation (it sits outside the
            // word characters), so no manual re-attachment is needed.
            var cleanWord = StripTrailingPunctuation(word);
            if (string.IsNullOrEmpty(cleanWord))
            {
                continue;
            }

            var candidateCodes = BeiderMorse.Encode(cleanWord);
            if (candidateCodes.Count == 0)
            {
                continue;
            }

            foreach (var entry in _encodedVocabulary)
            {
                // Already the correct spelling (case-insensitive) — nothing to do.
                if (string.Equals(cleanWord, entry.OriginalWord, StringComparison.OrdinalIgnoreCase))
                {
                    break;
                }

                var hasMatch = entry.PhoneticCodes.Any(dictCode => candidateCodes.Contains(dictCode));
                if (!hasMatch)
                {
                    continue;
                }

                // Word-boundary anchored, escaped regex replace (matches
                // VocabularyProcessor). Replace only the cleanWord that matched so
                // substrings of other words are left intact.
                var pattern = $@"\b{Regex.Escape(cleanWord)}\b";
                corrected = Regex.Replace(
                    corrected,
                    pattern,
                    entry.OriginalWord.Replace("$", "$$"),
                    // CultureInvariant so IgnoreCase folds ASCII consistently under
                    // any locale (e.g. the Turkish dotless-i) — matches the
                    // convention in VocabularyProcessor.cs.
                    RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
                LoggingService.Debug($"Phonetic match: '{word}' -> '{entry.OriginalWord}'");
                break; // Use first match.
            }
        }

        return corrected;
    }

    /// <summary>Strip trailing punctuation from a word, returning the clean word.</summary>
    private static string StripTrailingPunctuation(string word)
    {
        var end = word.Length;
        while (end > 0 && char.IsPunctuation(word[end - 1]))
        {
            end--;
        }
        return word[..end];
    }
}

/// <summary>
/// Thin C# wrapper over the shared Rust core's Beider-Morse phonetic encoder,
/// mirroring macOS <c>BeiderMorse</c>. Backed by <c>phonetic_encode</c> from the
/// UniFFI-generated <c>hyperwhisper_core.cs</c> binding.
/// </summary>
internal static class BeiderMorse
{
    /// <summary>
    /// Encode a word into its Beider-Morse phonetic representations.
    /// Returns an empty list for empty input.
    /// </summary>
    public static IReadOnlyList<string> Encode(string word)
    {
        if (string.IsNullOrEmpty(word))
        {
            return Array.Empty<string>();
        }
        return HyperwhisperCoreMethods.PhoneticEncode(word);
    }
}
